### DISCUSSION TITLE
[Question] Make device supervised by panel

### DISCUSSION THREAD

**Discussion by Alviunta** [taligentx/dscKeybusInterface#279] [dscKeybusInterface Discussion]
Great job!
Does anyone know if it is possible to get the panel to monitor the arduino connection?
As far as I understand, the modules are supervised by the panels through the code 0x11 (module supervision query) to which each module places a series of bits at a specific time to indicate presence and connection, so if a module is disconnected it is disconnected you can see this directly on the keypad.
As the panel can support up to 8 keypads, perhaps a slot could be used to monitor the aruino connection and ensure that it is working at all times. I tried to use panelData [0] == 0x11 and a timer to write 00 in the KeyPad 2 slot (slot 8 was already in use) but it didn't work out as expected and the panel showed a module error, I don't know if any other is necessary type of interaction with the panel.

I have a series of modules that can be of help to try to decipher more information and achieve what I am proposing, as soon as I have more time I will get to work on it:
- [Escort5580](https://www.dsc.com/index.php?n=products&o=view&id=18) telephone interface
- [IT100 bi-directional RS232](https://www.dsc.com/alarm-security-products/IT-100%20-%20PowerSeries%20Integration%20Module/22) module (used to make integrations with other systems)
- IP communicator (I don't know the model)
- 8-Hardwire Zone Expander [PC5108](https://www.dsc.com/index.php?n=Products&o=view&id=39)
- Power Supply Modules [PC5204](https://www.dsc.com/index.php?n=Products&o=view&id=41)

On the other hand, I have seen third-party IP communicators that connect to the KeyBus and wrap themselves in the [printer](https://www.dsc.com/alarm-security-products/PC5400%20-%20Printer%20Interface%20Module/44) slot (I suppose to explicitly obtain all the information or something like that). Does anyone have information on this? Perhaps this could make things easier for the future. If I get a module that works like this, I will gladly share what I find.

Is there any documentation on how the KeyBus protocol works? I looked in many places but I could find little, if someone can give me help I would be grateful.

I apologize for my bad English (Google Translate did most of the work) and I hope someone can help me!

**Comment by taligentx** [comment]
>Does anyone know if it is possible to get the panel to monitor the arduino connection?

Hi @Alviunta - this should be possible, I've considered it previously but it's been lower priority as I haven't seen requests for this feature (until yours) and it would use additional resources on the microcontrollers (mostly an issue for Arduino).  As you mentioned it would be a matter of deciding which type of module to emulate and then decode the Keybus signaling for it.  Emulating a keypad should be straightforward since the 0x11 command is decoded.

>I have a series of modules that can be of help to try to decipher more information and achieve what I am proposing, as soon as I have more time I will get to work on it

This would be great!  Getting more of the Keybus decoded is always a goal for this library.

>On the other hand, I have seen third-party IP communicators that connect to the KeyBus and wrap themselves in the printer slot

I haven't seen data for the PC5400 printer module, logs from the `KeybusReader` sketch would be handy if anyone has the module.  Ideally, a log with the module disconnected, then another log with the module connected.

>Is there any documentation on how the KeyBus protocol works?

[This thread on AVRFreaks](https://www.avrfreaks.net/forum/dsc-keybus-protocol?page=all) is where I first came across folks decoding the Keybus protocol.  The [dscKeybusPrintData.cpp](https://github.com/taligentx/dscKeybusInterface/blob/develop/src/dscKeybusPrintData.cpp) file documents the parts of the protocol that are currently decoded.

>I apologize for my bad English (Google Translate did most of the work) and I hope someone can help me!

No apologies needed, your post was very clear!

Thanks,
Nikhil

**Comment by Dilbert66** [comment]
Supervision is definitively doable but in my opinion is not really a useful addition since when you think about it, the main idea of this library is to monitor and control the alarm system and not the reverse so having the alarm system monitor the keypad emulation seems unecessary. Of course, I could be missing a scenario.  

Anyhow, I did implement supervision with my expander board emulation version of this library.  The firmware fully allowed the emulated expander boards to be supervised. I did not supervise the keypad since i found it unecessary but it would simply have needed the addition of setting a predefined unused slot as you noted.   You need to be able to respond to the cmd 11 directly in the ISR routine to send back the slot info.   Here's a link to the branch where you can view the implementation.   

https://github.com/taligentx/dscKeybusInterface/blob/expander/src/dscKeybusInterface.cpp

In my own version , I don't bother activating the supervision as I found it to be just very annoying since during testing/development needed the restart of the chip and the panel would always scream that a device is missing at that time!  

Nikhil is correct in noting that it will impact resource use on lower ram devices so would need to be a #define to limit this function to more powerful devices.   I am rewriting the expander version to be more modular and fit better within the new structure of Nikhil's library.

**Reply by Alviunta** [reply]
Hi @Dilbert66 @taligentx, sorry for the delay, I had some problems at home.

> I haven't seen data for the PC5400 printer module, logs from the KeybusReader sketch would be handy if anyone has the module. Ideally, a log with the module disconnected, then another log with the module connected.

Soon I will get to work on this aspect, when I have the logs I will share them.

> Supervision is definitively doable but in my opinion is not really a useful addition since when you think about it, the main idea of this library is to monitor and control the alarm system and not the reverse so having the alarm system monitor the keypad emulation seems unecessary. Of course, I could be missing a scenario.

I understand where you mean that module supervision may be unnecessary, however, since it is a security system that works at all times, I believe that a "bidirectional" supervision provides a plus in terms of ensuring correct operation.
I am going to pose a hypothetical scenario: a house has an alarm system with remote monitoring and a fire starts. If there are no smoke sensors, it is necessary to activate the corresponding panic button on the keyboard in order to generate a report to the monitoring station and call the fire department, with this library you could do that activation remotely, from a cell phone for example.
With this in mind, imagine what would happen if the module you think is connected is actually not (!!!!) I would think that the firemen are on their way and they will never arrive, because nobody ever called them (!!!!)
Possibly the previously mentioned communication problem could be detected in time with a bidirectional supervision, because as you may have already noticed, when there is a supervision error nothing works, and forces us to review and fix whatever is wrong, in the worst case cases, I will have to disconnect my arduino from the system, but at least I will not run into the problem in the middle of an emergency.
This is an extreme and unlikely scenario, but it is possible, and it is the kind of scenario for which I decide to have a security system. If I have no guarantee that it will work in these situations, why would I want it?!

> Anyhow, I did implement supervision with my expander board emulation version of this library. The firmware fully allowed the emulated expander boards to be supervised. I did not supervise the keypad since i found it unecessary but it would simply have needed the addition of setting a predefined unused slot as you noted. You need to be able to respond to the cmd 11 directly in the ISR routine to send back the slot info. Here's a link to the branch where you can view the implementation.

I looked at your implementation (briefly) and it seems to try to emulate "virtual" zones, it seems to me an excellent addition to this library but it goes beyond what I'm saying. My interest is to ensure that the device that I connect to the alarm system is working at all times. And that if at any time it fails, I find out at that moment and not in an undesirable situation.

> In my own version , I don't bother activating the supervision as I found it to be just very annoying since during testing/development needed the restart of the chip and the panel would always scream that a device is missing at that time!
> Nikhil is correct in noting that it will impact resource use on lower ram devices so would need to be a #define to limit this function to more powerful devices. I am rewriting the expander version to be more modular and fit better within the new structure of Nikhil's library.

When dealing with systems that must act in extreme and urgent situations, more supervision is always better, as far as I am concerned, I prefer to spend a little more money on a more powerful microcontroller instead of spending a lot of money on replacing stolen / burned things.
I fully agree to do it in a modular way, so that each user achieves the solution that best suits their needs, and I understand that during development it is annoying to have to restart the system, but thinking in the long term I would not like a scenario as the previous one happened. If I'm going to have it connected to my system I want to be 100% sure it will work when I need it.

I hope what I am trying to say is understood.

**Comment by taligentx** [comment]
>  I am rewriting the expander version to be more modular and fit better within the new structure of Nikhil's library.

Hi Alain, this is great to hear, integrating your expander code has been on my TODO list for too long but hopefully with the new structure it'll be easier to include different functionalities while keeping lower resource utilization for more basic uses.

I'm also looking at making writes to the panel more modular, the current code is based on writing as a keypad, but I think things like responding to supervision and emulating DSC modules could use a better structure.

**Reply by Dilbert66** [reply]
Hi Nikhil, I'm currently playing around on adding full LCD keypad emulation to the ESPHOME example and in doing this, I need to emulate full installer programming functions.  One of the main features is to be able to update multi character field entries such as phone numbers which have 32 characters.  This uses a mechanism of sending 0xA5 cmd on the 05/0a and receiving in response multiple 6E cmds of 4 bytes (holding 8 bcd numbers).  For 32 charactes that's 4 6E's.   For writing this data back, the keypad sends a AA cmd on the 05/0A and receives in response a 70 cmd which works the same as a 6e except for writing.  The keypad responds with multiple 70's with the data to be written back to the cmd section.   

Anyhow, this works similar to what I had done with the expander code but it's getting messy and duplicates really what the keypad write code does.  The ISR is code is getting too large.  So as you noted, a better algorithm is needed.    What I'm working on is a queueing mechanism where each device that needs to respond on a cmd, pushes it's message unto a queue. (or separate queues for each cmd)   This message will include things like the data to send, the data lenght and the cmd it should be sent on.  This would be stored in a structure.     The ISR routine would then check on each cmd for any pending queue items  for that cmd and process them.  This would keep the ISR routine small since it would be universal for all types and there would be no issue with missed cmds since they would be processed as FIFO events.   Also all event pre-processing would be done outside of the ISR keeping it as lean as possible.   A critical part is the planning of the queue structure and retrieval.   This is very similar to what I did with the VISTA panel firmware.

**Reply by Dilbert66** [reply]
Some cmds would not need a queue such as the expanders.  The cmd would just keep getting an updated field with fresh data.  So cmds which are request cmds such as 05/0a/e6.20 would be queued and response cmds would be handled by a case statement which would return the associated cmd data as well as clearing the original request cmd. Just thinking out loud.  Different ways of handling it.

**Comment by Dilbert66** [comment]
I agree Nikhil. A structure change would be benificial to allow these external components to be modular and to be able to enable/disable them easily. I've encountered some low IRAM issues with the ESP8266 when combined with the ESPHOME library which needed compile time defines to turn off the expander portion when not needed.  I'll wait to see what you come up with before working on recoding my end. One critical improvment will be to optimize IRAM useage.

**Comment by bennydiamond** [comment]
I recently made my own interface using this library and made a bunch of config changes to my panel. I have a "Module Supervision" trouble and I was wondering if it might be the arduino module being wrongly "identified" by the panel and triggering the trouble. 

I'm looking for confirmation that this library does not implement a concept of supervised module and thus could not be the cause of the "module supervision" trouble I am having.

**Reply by kricon** [reply]
Hi. This library as-is doesnt have any supervision reporting to panel nor panel can see it.
Your trouble is coming from some panel config changes - try to initialize module supervision reset by entering `*8 [installer_code] 902 #` and then after minute or so enter `*8 [installer_code] 903` to verify supervised modules and if trouble has gone away.

**Comment by kricon** [comment]
Adding supervision (assigning ESP as keypad slot) capability could be useful when using it with SMS-GSM or telegram/push examples - so when ESP goes offline/stuck you will get trouble light and know that there's something wrong with standalone ESP notifier and wouldn't get notified if alarm occurs.
